plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java

```diff
@@ -64,90 +60,100 @@ public PhoenixMergeSink(PhoenixClient phoenixClient, RemoteQueryModifier remoteQ
+    private static ConnectorPageSink createDeleteSink(
             ConnectorSession session,
-            List<Type> mergeRowIdFieldTypes,
-            PhoenixClient phoenixClient,
+            JdbcClient jdbcClient,
+            RemoteQueryModifier remoteQueryModifier,
             ConnectorPageSinkId pageSinkId,
-            RemoteQueryModifier remoteQueryModifier)
+            String schemaName,
+            String tableName,
+            JdbcColumnHandle mergeRowIdColumnHandle)
     {
-        checkArgument(mergeRowIdFieldNames.size() == mergeRowIdFieldTypes.size(), "Wrong merge row column, columns and types size not match");
-        JdbcOutputTableHandle deleteOutputTableHandle = new PhoenixOutputTableHandle(
-                schemaName,
-                tableName,
-                mergeRowIdFieldNames,
-                mergeRowIdFieldTypes,
-                Optional.empty(),
-                Optional.empty());
+        RowType mergeRowIdType = (RowType) mergeRowIdColumnHandle.getColumnType();
 
-        return new DeleteSink(session, deleteOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
+        ImmutableList.Builder<String> mergeRowIdFieldNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> mergeRowIdFieldTypesBuilder = ImmutableList.builder();
+        for (RowType.Field field : mergeRowIdType.getFields()) {
+            checkArgument(field.getName().isPresent(), "Merge row id column field must have name");
+            mergeRowIdFieldNamesBuilder.add(getEscapedArgument(field.getName().get()));
+            mergeRowIdFieldTypesBuilder.add(field.getType());
+        }
+        List<String> mergeRowIdFieldNames = mergeRowIdFieldNamesBuilder.build();
+
+        String deleteSql = getDeleteSql(schemaName, tableName, mergeRowIdFieldNames);
+        return new JdbcPageSink(
+                session,
+                new PhoenixOutputTableHandle(
+                        schemaName,
+                        tableName,
+                        mergeRowIdFieldNames,
+                        mergeRowIdFieldTypesBuilder.build(),
+                        Optional.empty(),
+                        Optional.empty()),
+                jdbcClient,
+                pageSinkId,
+                remoteQueryModifier,
+                Optional.of(deleteSql));
     }
 
-    private class DeleteSink
-            extends JdbcPageSink
+    private static String getDeleteSql(
+            String schemaName,
+            String tableName,
+            List<String> mergeRowIdFieldNames)
     {
-        public DeleteSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
-        {
-            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
-        }
+        List<String> conjuncts = mergeRowIdFieldNames.stream()
+                .map(name -> name + " = ? ")
```

Comment:
I wonder if we should not use `columnWriters` from `JdbcPageSink` here. Notice that handling parameters is not trivial and I think we should somehow reuse it.

I need to think more how... any ideas?

Reply:
I am trying to use the `DefaultQueryBuilder#prepareDeleteQuery`, this is the code but not sure I am in the right direction.

Reply:
Hi, could you please help to have a look at the try. I am going to refactor the `Phoenix` until we meet the agreement, then I will modify the first commit of the #23034

Reply:
I think it is looking nice.